### DISCUSSION TITLE
Support for AWS_MSK_IAM

### DISCUSSION THREAD

**Discussion by mbartusiak** [edenhill/librdkafka#3385] [General]
Hello

AWS announced support for a new SASL mechanism called AWS_MSK_IAM which enables to handle both authentication and authorisation with AWS IAM. Is there a chance to add support for that mechanism to librdkafka?

Cheers
Michal

**Comment by edenhill** [comment]
Is that its own mechanism or on top of OAUTHBEARER?

**Reply by mbartusiak** [reply]
This is their custom SASL mechanism. Sadly, they only provide a Java library.
More on this:
https://aws.amazon.com/blogs/big-data/securing-apache-kafka-is-easy-and-familiar-with-iam-access-control-for-amazon-msk/
> The feature only works out of the box for Java-based clients. If you’re interested in using IAM authentication with Apache Kafka clients written in languages other than Java, see the following [GitHub repo]( https://github.com/aws/aws-msk-iam-auth). At a high level, this involves implementing a SASL mechanism that uses the AWS SDK to fetch client credentials and sends a very specific JSON authentication payload to the MSK cluster, which then automatically handles access control.

**Reply by rolandyoung** [reply]
Given that this is for MSK (Amazon Managed Streaming For Apache Kafka) and that librdkafka doesn't have direct support for Apache Streams, is there much of a use case for supporting it?

**Reply by mbartusiak** [reply]
AWS MSK is just a managed kafka cluster, utilising streams in apps is optional. You can still use basic producers and consumers

**Reply by sayantacC** [reply]
I am with AWS and a contributor to [aws-msk-iam-auth](https://github.com/aws/aws-msk-iam-auth) repo , the jvm based plugin for integrating AWS IAM authentication and authorization as a SASL mechanism with Kafka. We would love to engage with you about adding support for the AWS IAM SASL mechanism to librdkafka.

 Based on initial exploration, we have identified two possible approaches:
 1. Write a librdkafka plugin to implement and register an additional sasl provider. However, as far as I could tell, librdkafka's plugin architecture will have to be extended to allow a plugin to implement and set the sasl provider based on the configured mechanism.

 2. Implement AWS_MSK_IAM as another sasl mechanism within librdkafka. 

 We would love to learn if there are other possible approaches and your thoughts on which one to take. 
 Also, any suggestions on next steps would be great.

**Reply by twmb** [reply]
For potential cross-language implementation reference, I've added support for AWS_MSK_IAM in my [franz-go](https://github.com/twmb/franz-go/blob/master/CHANGELOG.md#v073) client. I've tested it against a scrappy cluster setup, but I'd be happy to have broader testing (and I'm planning to turn down my cluster imminently).

I think that implementing the bare minimum of this sasl (similar to how I did it) would not require too much C code.

**Reply by sayantacC** [reply]
I created an [issue](https://github.com/edenhill/librdkafka/issues/3402) for this feature request.
